File: src/utils/recordImages.ts
/**
 * Image detection for the rich JSON record preview.
 *
 * Records reference images two ways, and both should render a thumbnail in the
 * structured field view:
 *
 *   1. A direct HTTP(S) URL in a string field (e.g. arena mirror's `image.src`,
 *      an external `thumbnail`, etc.). Rendered straight from the URL.
 *   2. An AT Protocol blob with an image `mimeType` (e.g. a profile avatar, or
 *      arena mirror's `image.blob`). Rendered via `com.atproto.sync.getBlob`
 *      against the owning PDS, so it needs the repo's DID + PDS endpoint.
 *
 * These are pure helpers (no React, no fetch) so both the preview component and
 * any future caller share one definition of "is this field an image".
 */

/** Path extensions we're willing to render inline as an image. */
const IMAGE_EXT_RE =
  /\.(jpe?g|png|gif|webp|avif|svg|bmp|ico|apng|jfif|heic|heif|tiff?)$/i;

/**
 * If `value` is a string that points at a renderable image over HTTP(S),
 * return the URL; otherwise null. The extension is checked against the URL's
 * pathname only, so query strings (arena appends `?<timestamp>`) and fragments
 * don't defeat the match. The raw URL is returned un-sanitized — callers pass
 * it through `sanitizeUrl` before putting it in the DOM.
 */
export function imageUrlFromValue(value: unknown): string | null {
  if (typeof value !== 'string') return null;
  const s = value.trim();
  if (!/^https?:\/\//i.test(s)) return null;
  let pathname: string;
  try {
    pathname = new URL(s).pathname;
  } catch {
    return null;
  }
  return IMAGE_EXT_RE.test(pathname) ? s : null;
}

export type ImageBlobRef = { cid: string; mimeType: string };

/**
 * If `value` is an AT Protocol blob describing an image, return its CID +
 * mimeType; otherwise null. Handles both the current blob shape
 * (`{ $type: 'blob', ref: { $link }, mimeType }`) and the legacy inline-CID
 * shape (`{ cid, mimeType }`). Non-image blobs (video, etc.) return null so we
 * never try to render them as `<img>`.
 */
export function imageBlobFromValue(value: unknown): ImageBlobRef | null {
  if (!value || typeof value !== 'object') return null;
  const obj = value as Record<string, unknown>;
  const mimeType = typeof obj.mimeType === 'string' ? obj.mimeType : null;
  if (!mimeType || !mimeType.startsWith('image/')) return null;

  const ref = obj.ref;
  if (ref && typeof ref === 'object') {
    const link = (ref as Record<string, unknown>)['$link'];
    if (typeof link === 'string' && link) return { cid: link, mimeType };
  }
  // Legacy shape: the CID sits directly on the blob.
  if (typeof obj.cid === 'string' && obj.cid) return { cid: obj.cid, mimeType };
  return null;
}

/**
 * Build the public `com.atproto.sync.getBlob` URL that serves a blob's bytes
 * straight from the owning PDS.
 */
export function getBlobUrl(pds: string, did: string, cid: string): string {
  const params = new URLSearchParams({ did, cid });
  return `${pds.replace(/\/$/, '')}/xrpc/com.atproto.sync.getBlob?${params}`;
}

/** Pull the DID authority out of an `at://did:.../…` URI (null if not a DID). */
export function didFromAtUri(uri: string): string | null {
  const match = /^at:\/\/(did:[^/]+)/.exec(uri);
  return match ? match[1] : null;
}

File: ios/Packages/AturiCore/Sources/AturiCore/Records/RecordImages.swift
import Foundation

/// An AT Protocol blob reference that describes an image.
public struct ImageBlobRef: Hashable, Sendable {
    public var cid: String
    public var mimeType: String

    public init(cid: String, mimeType: String) {
        self.cid = cid
        self.mimeType = mimeType
    }
}

/// Image detection for the structured record view. Port of
/// `src/utils/recordImages.ts`.
///
/// Records reference images two ways and both should render a thumbnail:
/// a direct http(s) URL in a string field, and a blob with an image
/// `mimeType`, which is served by `com.atproto.sync.getBlob` on the owning
/// PDS and therefore needs the repo's DID and PDS endpoint.
public enum RecordImages {
    /// Path extensions worth rendering inline. Checked against the URL's
    /// path only, so query strings and fragments cannot defeat the match.
    private static let imageExtensionPattern =
        #/\.(jpe?g|png|gif|webp|avif|svg|bmp|ico|apng|jfif|heic|heif|tiff?)$/#.ignoresCase()

    private static let httpSchemePattern = #/^https?:\/\//#.ignoresCase()

    /// The trimmed string when `value` is an http(s) URL whose path ends in
    /// an image extension; nil otherwise. Returned unsanitised, exactly as
    /// the field spelled it.
    public static func imageUrlFromValue(_ value: JSONValue?) -> String? {
        guard let s = value?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines) else { return nil }
        guard s.prefixMatch(of: httpSchemePattern) != nil else { return nil }
        guard let path = URLComponents(string: s)?.path
            ?? URLComponents(string: s.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? s)?.path
        else { return nil }
        return path.firstMatch(of: imageExtensionPattern) != nil ? s : nil
    }

    /// The CID and mimeType when `value` is a blob describing an image, in
    /// the current shape (`{ $type: 'blob', ref: { $link }, mimeType }`) or
    /// the legacy inline-CID shape (`{ cid, mimeType }`). Non-image blobs
    /// (video, ...) return nil.
    public static func imageBlobFromValue(_ value: JSONValue?) -> ImageBlobRef? {
        guard let object = value?.objectValue else { return nil }
        guard let mimeType = object["mimeType"]?.stringValue, mimeType.hasPrefix("image/") else { return nil }
        if let link = object["ref"]?["$link"]?.stringValue, !link.isEmpty {
            return ImageBlobRef(cid: link, mimeType: mimeType)
        }
        if let cid = object["cid"]?.stringValue, !cid.isEmpty {
            return ImageBlobRef(cid: cid, mimeType: mimeType)
        }
        return nil
    }

    /// The public `com.atproto.sync.getBlob` URL that serves a blob's bytes
    /// from the owning PDS. A trailing slash on `pds` is dropped.
    public static func getBlobUrl(pds: String, did: String, cid: String) -> String {
        let base = pds.hasSuffix("/") ? String(pds.dropLast()) : pds
        return base + "/xrpc/com.atproto.sync.getBlob?did=" + URIEncoding.encodeComponent(did)
            + "&cid=" + URIEncoding.encodeComponent(cid)
    }

    /// The DID authority of an `at://did:.../...` URI; nil when the
    /// authority is a handle or the input is not an AT URI.
    public static func didFromAtUri(_ uri: String) -> String? {
        guard let match = uri.prefixMatch(of: didAuthorityPattern) else { return nil }
        return String(match.output.1)
    }

    private static let didAuthorityPattern = #/^at:\/\/(did:[^/]+)/#
}
